zcash_client_backend/src/fees/zip317.rs

```diff
@@ -351,13 +350,149 @@ mod tests {
             ),
             #[cfg(feature = "orchard")]
             &orchard_fees::EmptyBundleView,
+            &DustOutputPolicy::new(DustAction::AllowDustChange, Some(NonNegativeAmount::ZERO)),
```

Comment:
Why was the `change_with_transparent_payments` test altered to use this dust output policy (instead of `DustOutputPolicy::default()`)? AFAICT the test is not transparent-only (it has a Sapling input), and therefore the change to the fee code should not have affected it.

If this isn't a necessary change, revert it. If we don't have a test with `DustAction::AllowDustChange`, then we should add that separately.

Reply:
The behaviour intentionally changes in the situation of the original test. Specifically, one of the dummy outputs that was originally present would become a change output instead, because we provided a change memo. (This would be observable to someone with the corresponding internal IVK, although not otherwise.) However, the default dust policy disallows creating a transaction with zero-valued change (the default dust threshold is the ZIP 317 marginal fee of 5000 zat), and so the test would fail without changing the dust policy.

I don't know whether this is the intent of the default policy, but it is strictly speaking what we asked for: we specified a change memo and said to reject if the change is less than the marginal fee. So what I did to make the test pass is explicitly specify a policy that allowed this case (which might not be the correct solution).

In more detail:

The original test has a Sapling input of 55000 zat (so, the flows are not fully transparent), and a transparent output of 40000 zat. The selected change pool is Sapling. There is no explicitly specified Sapling output (and the `bundle_required` setting of the builder defaults to false), but there is at least one Sapling input and so the outputs are padded to 2; [this is the intended behaviour of the Sapling builder](https://github.com/zcash/sapling-crypto/blob/d23547f0d3c753b706c6a4fd64456855e153d6b7/src/builder.rs#L36..L37). There is no Sapling change output. Therefore the ZIP 317 fee is, correctly, 15000 zat because we have 3 logical actions: two due to Sapling (max of one input and two dummy outputs), and one transparent output.

Before the latest changes from our pairing, the PR was incorrect because it was calculating the fee as 20000 zat.

The new behaviour is that we create a zero-valued change output with `change_memo`. Because of the output padding, this makes no difference to the fee of 15000 zat, and so the balance is calculated as 55000 - 40000 - 15000 = 0, i.e. we are in case 2, but the flows are not transparent so this becomes case 3b. However, the default dust policy disallows the change output. Therefore, we reject the transaction as having insufficient funds.

I'm open to either:
* renaming the test;
* having this test explicitly fail and adding another test that allows zero-valued change;
* changing the interpretation of the default dust policy to allow this case. (Should it also allow change less than the marginal fee in general if a change memo is given, or only zero-valued change?)

Reply:
An argument for allowing zero-valued change but not allowing other change values less than the threshold, is that we will never need to spend a zero-valued note (and in principle shouldn't be keeping track of its witness, although I don't know whether that is currently the case). And this case may occur in practice for the intended "send max" feature in Zashi.

Reply:
> Because of the output padding, this makes no difference to the fee of 15000 zat, and so the balance is calculated as 55000 - 40000 - 15000 = 0, i.e. we are in case 2, **but the flows are not transparent so this becomes case 3b**.

This is the bit of logic that was missing when I originally wrote this comment in the previous PR. I think this test is therefore fine as-is now.

Reply:
I've implemented this now; please re-review.